### ISSUE TITLE
pan/zoom limits

### ISSUE THREAD

**Issue by retrotails** [ArturKovacs/emulsion#122]
right now, you can completely pan the image off screen and lose it, zoom out so far it's less than a pixel, or zoom in such that one pixel is many times larger than the window. I'd recommend something like:
don't let the top edge of the image go below 10% of the window height, same for all edges
zoom in maximum is 32x32 pixels or native resolution, whichever is more zoom
zoom out maximum is either 10% window size or the native image resolution, whichever is zoomed out more
I'm not any good with rust but I gave it a shot, here's the very first step which limits the center of the image to the corners of the screen, in picture_widget.rs:
```rust
fn min<T:PartialOrd>(a:T,b:T)->T { if a<b{a}else{b}}
fn max<T:PartialOrd>(a:T,b:T)->T { if a>b{a}else{b}}
```
```rust
if borrowed.panning {
	let delta = event.cursor_pos - borrowed.last_mouse_pos;
	borrowed.scaling = ScalingMode::Fixed;
	borrowed.update_scaling_buttons();
	borrowed.img_pos += delta;
	
	borrowed.img_pos = LogicalVector::new(
		max(borrowed.img_pos.vec.x, 0.0),
		max(borrowed.img_pos.vec.y, 0.0)
	);
	borrowed.img_pos = LogicalVector::new(
		min(borrowed.img_pos.vec.x, borrowed.drawn_bounds.size.vec.x),
		min(borrowed.img_pos.vec.y, borrowed.drawn_bounds.size.vec.y)
	);
	
	borrowed.render_validity.invalidate();
}
```
it's very crude and doesn't prevent you from moving out of bounds by zooming. if nobody else wants to tackle it I might try actually learning the code and implementing it.

also I think it might make sense to reset the pan/zoom when you switch images, though that's just my preference.

**Comment by ArturKovacs** [comment]
It is annoying when one loses the image and feels that there's no simple way to get it back.

However there's always a simple way to get the image back. By triggering "Best fit" or "Fit to borders" (<kbd>F</kbd> and <kbd>E</kbd> by default) the image size and position will be reset. Therefore I'm not sure if these things should be constrained. I really like that Emulsion respects the user's decision and input. An alternative to constraints would be to highlight the "Best fit" button for example when trying to zoom or pan at the extremes.

**Comment by retrotails** [comment]
okay so I can see how something a bit specific like I recommended might not fit the project, but maybe have something simple like this to prevent the image from going completely out of bounds.
https://github.com/retrotails/emulsion
oh also this fixes what I believe is a typo, pyhs->phys

**Comment by ArturKovacs** [comment]
Resolved by #126
